# 2022-04-20, test traefik and gitlab
# sudo touch /data/acme.json && chmod 600 /data/acme.json
# docker network create traefik-public
# domain name: wode.com
version: "3.7"
services:

  traefik:
    image: traefik:v2.6.3
    container_name: traefik
    restart: always
    networks:
      - traefik-public
    ports: 
      - "80:80"
      - "443:443"
    environment: 
      - "TZ=Asia/Shanghai"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - /var/log/traefik/:/logs/
      - /data/acme.json:/acme.json
    command:
      - "--providers.docker=true"
      - "--providers.docker.swarmMode=false"
      #- "--providers.docker.constraints=LabelRegx(`traefik.constraint-label-stack`, `git.+`)"
      - "--providers.docker.exposedbydefault=false"
      - "--providers.file.watch=true"
      - "--entryPoints.http.address=:80"
      - "--entryPoints.https.address=:443"
      - "--entryPoints.http.forwardedHeaders.trustedIPs=0.0.0.0/0"
      - "--entryPoints.traefik.address=:8080"
      - "--log=true"
      - "--log.level=INFO"
      - "--log.filePath=/logs/traefik.log"
      - "--api=true"
      - "--api.insecure=true"
      - "--ping=true"
      - "--accessLog=true"
      - "--accesslog.filePath=/logs/traefik-access.log"
      - "--accesslog.filters.statuscodes=200,300-302"
      - "--accesslog.filters.minduration=200ms"
      - "--accesslog.filters.retryattempts"
      - "--accesslog.bufferingSize=200"
      - "--accesslog.fields.headers.names.ClientAddr=keep"

      - "--certificatesResolvers.lets-encrypt.acme.httpChallenge=true"
      - "--certificatesResolvers.lets-encrypt.acme.storage=acme.json"
      - "--certificatesResolvers.lets-encrypt.acme.email=xxxxxx@163.com"
      - "--certificatesResolvers.lets-encrypt.acme.httpChallenge.entrypoint=http"


    deploy:
      resources:
        limits:
          cpus: '0.50'
          memory: 500M
        reservations:
            memory: 200M

    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=traefik-public"
      - "traefik.constraint-label=traefik_public_tag"
      - "traefik.http.routers.stack-proxy-http.rule=Host(`traefik.wode.com`)"
      - "traefik.http.routers.stack-proxy-http.entrypoints=https"
      - "traefik.http.services.stack-porxy.loadbalancer.server.port=8080"

      # config below domains to https
      - "traefik.http.routers.stack-proxy-http.tls.certresolver=lets-encrypt"
      - "traefik.http.routers.stack-proxy-http.tls.domains[0].main=traefik.wode.com"
      - "traefik.http.routers.stack-proxy-http.tls.domains[0].sans=wode.com"

      # HTTP REDIRECT
      - "traefik.http.middlewares.redirect-to-https.redirectscheme.scheme=https"
      - "traefik.http.routers.redirect-https.rule=hostregexp(`{host:.+}`)"
      - "traefik.http.routers.redirect-https.entrypoints=http"
      - "traefik.http.routers.redirect-https.middlewares=redirect-to-https"


  gitlab:
    image: gitlab/gitlab-ce:13.10.0-ce.0
    container_name: my-gitlab
    restart: always
    hostname: git.wode.com
    networks:
      - traefik-public
    ports:
      - "10081:80"
      - "10083:443"
      - "9922:22"
    environment:
      - TZ=Asia/Shanghai
      - DEBUG=false
      - GITLAB_OMNIBUS_CONFIG=external_url 'http://git.wode.com';gitlab_rails['time_zone']='Asia/Shanghai';
  
    volumes:
      - "/data/gitlab/etc:/etc/gitlab"
      - "/data/gitlab/log:/var/log/gitlab"
      - "/data/gitlab/opt:/var/opt/gitlab"

    deploy:
      resources:
        limits:
          cpus: '4.0'
          memory: 8000M
        reservations:
            memory: 200M
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=traefik-public"
      - "traefik.http.routers.gitlab_insecure.entrypoints=http"
      - "traefik.http.routers.gitlab_insecure.rule=Host(`git.wode.com`)"
      - "traefik.http.routers.gitlab_insecure.middlewares=redirect-to-https@docker"

      - "traefik.http.routers.gitlab.tls=true"
      - "traefik.http.routers.gitlab.entrypoints=https"
      - "traefik.http.routers.gitlab.rule=Host(`git.wode.com`)"
      - "traefik.http.routers.gitlab.tls.certresolver=lets-encrypt"
      - "traefik.http.routers.gitlab.service=gitlab_service"
      - "traefik.http.services.gitlab_service.loadbalancer.server.port=80"
      

  tomcat:
    image: tomcat:8.5.31-jre8
    restart: always
    networks:
      - traefik-public
    ports:
      - "10080:8080"
    labels:
      - "traefik.enable=true"
      - "traefik.docker.network=traefik-public"
      - "traefik.http.routers.tomcat-frontend-http.entrypoints=https"
      - "traefik.http.routers.tomcat-frontend-http.rule=Host(`tomcat.wode.com`)"
      - "traefik.http.services.tomcat-frontend.loadbalancer.server.port=8080"
      - "traefik.http.routers.tomcat-frontend-http.tls.certresolver=lets-encrypt"
      

  portainer:
    image: portainer/portainer
    container_name: my-portainer
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /data/portainer:/data
    networks:
      - traefik-public
    labels:
      - traefik.enable=true
      - "traefik.docker.network=traefik-public"
      - traefik.http.routers.portainer_insecure.entrypoints=http
      - traefik.http.routers.portainer_insecure.rule=Host(`portainer.wode.com`)
      - traefik.http.routers.portainer_insecure.middlewares=redirect-to-https@docker

      - traefik.http.routers.portainer.entrypoints=https
      - traefik.http.routers.portainer.rule=Host(`portainer.wode.com`)
      - traefik.http.routers.portainer.tls.certresolver=lets-encrypt
      - traefik.http.services.portainer.loadbalancer.server.port=9000

networks:
  traefik-public:
    external: true

